stellargraph/core/element_data.py

```diff
@@ -303,74 +303,117 @@ def _numpyise(d, dtype):
     return {k: np.array(v, dtype=dtype) for k, v in d.items()}
 
 
+class AdjList:
+    def __init__(self, init_index):
+        self.init_index = init_index
+        self.undirected = self.ins = self.outs = None
+
+    def force_init(self):
+        self.undirected = self.init_index(ins=True, outs=True)
+        self.ins = self.init_index(ins=True, outs=False)
+        self.outs = self.init_index(ins=False, outs=True)
+
+    def lookup(self, *, ins, outs):
+        if ins and outs:
+            if self.undirected is None:
+                self.undirected = self.init_index(ins=True, outs=True)
+            return self.undirected
+        if ins:
+            if self.ins is None:
+                self.ins = self.init_index(ins=True, outs=False)
+            return self.ins
+        if outs:
+            if self.outs is None:
+                self.outs = self.init_index(ins=False, outs=True)
+            return self.outs
+
+        raise ValueError(
+            "expected at least one of 'ins' or 'outs' to be True, found neither"
+        )
+
+
 class EdgeData(ElementData):
     """
     Args:
         shared (pandas DataFrame): information for the edges
         type_starts (list of tuple of type name, int): the starting iloc of the edges of each type within ``shared``
+        nodes (NodeData): node data containing node types and IDs
     """
 
     _SHARED_REQUIRED_COLUMNS = [SOURCE, TARGET, WEIGHT]
 
-    def __init__(self, shared, type_starts):
+    def __init__(self, shared, type_starts, nodes):
         super().__init__(shared, type_starts)
 
         # cache these columns to avoid having to do more method and dict look-ups
         self.sources = self._column(SOURCE)
         self.targets = self._column(TARGET)
         self.weights = self._column(WEIGHT)
 
+        self.source_types = nodes.type_of_iloc(self.sources)
```

Comment:
I think this is going to noticeably increase memory usage, both because:

1. it's using type names not type ilocs
2. it's two extra O(num edges) arrays that aren't shared with anything (i.e. they're new) but are stored permanently

The first can be fixed with:

```suggestion
        self.source_types = nodes.type_ilocs[self.sources]
```

but the second is more problematic. Maybe these could be computed lazily too? (And/or, not even stored, just used transiently for initialising the adjacency lists?)

I think the benchmarks in the PR description aren't up to date with the current state? Maybe some new benchmarks could be reported, including the `test_allocation_benchmark` ones (to try to capture this addition)?

Reply:
Good point, yeah I don't think we need to store these at all actually. I'll run some new benchmarks afterwards